### PR TITLE
Add SNI support to client and server sockets.

### PR DESCRIPTION
**Pull request by nathan-hoad**
So to be clear, I don't consider this complete just yet, but I'm looking for some feedback as I'm very new to Nim. This is in fact the first code I've written, so I have a few noobish questions...

- Are there forward declarations for types? I had to move the definition of `TIpAddress` to before `connect`, so that I could use it.
- `RootEffect`, is this a "base" effect? I couldn't compile the code without adding it to the `connect` method definition, and it feels kind of bad.
- The interface on the callback functions is quite poor to use and involves casts, some to deal with net.SslContext being a distinct SslCtx. Is there a way to make this any nicer to use? (See below example)
- I've got a pretty loose set of integers that should be used for the return codes of the servername callback function. I think an enum would be nice, but I don't know how to enforce it and then mutate it back into an integer for OpenSSL.
- I haven't updated the ssl or sockets modules, as they are deprecated.

In general, how does this code seem, style and idiom wise? Happy to take any criticisms/change anything to get it accepted. Once the above issues have been resolved, I'll resubmit with documentation.

```
import net
import openssl

var s = newSocket()
var ctx = newContext(protVersion=protTLSv1, certFile="cert.pem", keyFile="key.pem")
ctx.wrapSocket(s)

s.bindAddr(Port(8081), "127.0.0.1")

var b = true

proc myfunc(ssl: SslPtr, cb_id: int, arg: pointer): int =
    var g = SSL_get_servername(ssl)
    if not isNil(g):
        echo "server name is ", g

    var c = cast[ptr bool](arg)

    echo "modifying arg value ", c[], " ", b
    c[] = false
    echo "modifed arg value? ", c[], " ", b

    return SSL_TLSEXT_ERR_NOACK

# FIXME: can these casts to SslCtx be avoided?
echo "setting callback", SSL_CTX_set_tlsext_servername_callback(cast[SslCtx](ctx), cast[PFunction](myfunc))
echo "setting callback ", SSL_CTX_set_tlsext_servername_arg(cast[SslCtx](ctx), addr(b))

while true:
    s.listen()

    var c = newSocket()
    s.accept(c)

    echo "accepted client"
```

### CODE REVIEW

File: lib/pure/net.nim [hunk 1]
```diff
@@ -511,6 +527,13 @@ proc connect*(socket: Socket, address: string, port = Port(0),
   
   when defined(ssl):
     if socket.isSSL:
+      try:
+        discard parseIpAddress(address)
+      except ValueError:
+        # Discard result in case OpenSSL version doesn't support SNI, or we're
+        # not using TLSv1+
+        discard SSL_set_tlsext_host_name(socket.sslHandle, address)
```
Comment (dom96) [inline lib/pure/net.nim]: I don't really like the way you are detecting whether ``address`` is an IP address or not. Is there no simpler way to do this?

**Comment by dom96** [comment]
You shouldn't be casting, that's unsafe. Use an object conversion: ``ctx.SslCtx``.

Your example is very low-level. It would be nice if you could write a high-level interface for this. I'm not an expert in SSL (and apart from what I just read on Wikipedia I have no idea what SNI is), so i'm not sure what the API for it should look like. Ideally we should avoid the callbacks.

File: lib/pure/net.nim (hunk 1 above)
Comment (nathan-hoad) [inline lib/pure/net.nim]: I can't think of one. It's against the RFC to have IP addresses in the server name (RFC 3546, section 3.1):

    Literal IPv4 and IPv6 addresses are not permitted in "HostName".

Happy for suggestions however! :)

Comment (dom96) [inline lib/pure/net.nim]: Please create a new proc and call it "isIpAddress". Just to make it a bit clearer about what you are actually checking there.

**Comment by nathan-hoad** [comment]
Sure! I'll fix up the casting, object conversion looks much nicer. I could and write a higher-level interface for this, but the callbacks are basically a necessity if you're implementing servers that need SNI.

As an explanation for SNI, Apache HTTP and NGINX (as well as lots of other HTTP servers, I'm sure) provide what's known as "virtual hosting", where you can have the HTTP server listen on one port, yet multiple services can run on it. For unencrypted HTTP, it's easy to identify the service that should handle the request by the `Host` header, but for HTTPS you can't do this as you can't observe the host header until you've negotiated the SSL connection. This is the problem that SNI solves - by having the client specify "This is the server I'm trying to connect to", it lets the server provide the correct certificate, and the correct service.

An example of this is "https://gmail.com" versus "https://google.com". Both domains live on the same IP addresses, but if you try to connect to "https://gmail.com" without SNI support, you'll get a certificate for "google.com" (i.e. an invalid certificate).

So the reason for the callbacks is to allow people to implement their own virtual server solution - it is not required for normal SNI client support, and quite specialised to virtual servers. I think that any higher-level interface would go ignored in favour of people rolling their own. A rudimentary example might be something like...

```
proc get_server(ssl: SslPtr, cb_id: int, arg: pointer): int =
    var hostname = SSL_get_servername(ssl)
    if isNil(hostname):
        echo "we don't have a hostname! bailing"
        return SSL_TLSEXT_ERR_NOACK

    var server = lookupServer(hostname)
    if isNil(server):
        echo "could not find suitable server"
        return SSL_TLSEXT_ERR_NOACK

    # set the context for this connection so the client receives the correct certificate
    if not isNil(server.sslCtx)
        echo "setting context for connection to ", server

        # this function isn't actually exposed in Nim yet, but I will submit another pull request with that later
        SSL_set_SSL_CTX(ssl, server.sslCtx)
    return SSL_TLSEXT_ERR_OK
```

This is a barebones implementation of how NGINX does it, but of course it could be done in lots of ways and everyone likes to roll their own. It also makes assumptions about what a "server" is, and is much closer to a "business-logic" kind of solution, which is not really what the `openssl` module is about I think.

**Comment by nathan-hoad** [comment]
Ah, how do I handle backwards compatibility/missing symbols? For example `SSL_get_servername` shouldn't be available if the dynlib doesn't support it, e.g. me compiling against an old version of openssl...

```
[Linking]
/home/nathan/Nim/nimcache/compiler_sslserver.o: In function `myfunc_142028':
compiler_sslserver.c:(.text+0x40c): undefined reference to `SSL_get_servername'
collect2: error: ld returned 1 exit status
```

I understand that handling it at runtime might be difficult, but at compile time at least should be an easy case to cover.

Sorry for all the questions! I could add a new define for it, but I'm not sure of the policy on adding them for these sorts of things :) I'm planning on submitting a lot more TLS related support in the future too that will need the same sort of thing (e.g. `TLS_v1_1_method()` and `TLV_v1_2_method()`) so it will be good information to have.

**Comment by Varriount** [comment]
This is currently a limitation of the builtin dll import pragmas - It's been suggested (well, I've suggested) that the body of a dll-imported procedure be used when the symbol can't be loaded, but so far this hasn't been implemented.

Your best bet for now is to use the dynlib module to manually search and load symbols from the openssl library, and handle the failures that way.

**Comment by nathan-hoad** [comment]
Hmm, okay, I can give that a go. Is there any precedent for doing this? I can look at doing the entire module or just for the relevant symbols, but it may be quite a bit of work and kinda ugly.

**Comment by Araq** [comment]
> *For example SSL_get_servername shouldn't be available if the dynlib doesn't support it, e.g. me compiling against an old version of openssl.*

Er, I think that's not the best example for this feature. Do not support or use old versions of openssl.

**Comment by dom96** [comment]
I think that perhaps dead code elimination could be a solution to these problems. As long as the user's code does not use those procedures they will be removed when ``--deadCodeElim:on`` is specified. This can also be specified using a pragma in a module.

**Comment by nathan-hoad** [comment]
@Araq Sure, I'm happy to not support old versions of OpenSSL - the version constant definitely supports old versions though, so I'll update that to remove the older versions that don't support these features.

@dom96 That's a good point! openssl already has a pragma on the module for this, so this should be fine I think.

**Comment by Araq** [comment]
@dom96 ping

**Pull request merged**